### PR TITLE
kube-aggregator: correctly use client-go TLS cache with custom dialer

### LINKED ISSUE #117250: Global transport cache grows unbounded in kube-apiserver process with each sync of APIServiceRegistrationController.
### What happened?

Initially observed gradual, linear growth in kube-apiserver's working set size (`container_memory_working_set_bytes`) over the course of a week in an OpenShift 4.13.0-rc.2 (downstream of Kubernetes v1.26.2) cluster that could not be reproduced on OpenShift 4.12.0 (Kubernetes v1.25.4).

A heap profile shows:

```
File: kube-apiserver
Type: inuse_space
Time: Apr 12, 2023 at 12:41pm (EDT)
Active filters:
   focus=APIServiceRegistrationController
Showing nodes accounting for 193MB, 15.97% of 1208.72MB total
Dropped 4 nodes (cum <= 6.04MB)
      flat  flat%   sum%        cum   cum%
         0     0%     0%   194.01MB 16.05%  k8s.io/apimachinery/pkg/util/wait.BackoffUntil
         0     0%     0%   194.01MB 16.05%  k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1
         0     0%     0%   194.01MB 16.05%  k8s.io/apimachinery/pkg/util/wait.JitterUntil
         0     0%     0%   194.01MB 16.05%  k8s.io/apimachinery/pkg/util/wait.Until
         0     0%     0%   194.01MB 16.05%  k8s.io/kube-aggregator/pkg/apiserver.(*APIAggregator).AddAPIService
         0     0%     0%   194.01MB 16.05%  k8s.io/kube-aggregator/pkg/apiserver.(*APIServiceRegistrationController).processNextWorkItem
         0     0%     0%   194.01MB 16.05%  k8s.io/kube-aggregator/pkg/apiserver.(*APIServiceRegistrationController).runWorker
         0     0%     0%   194.01MB 16.05%  k8s.io/kube-aggregator/pkg/apiserver.(*APIServiceRegistrationController).sync
       1MB 0.083% 0.083%      193MB 15.97%  k8s.io/kube-aggregator/pkg/apiserver.(*proxyHandler).updateAPIService
         0     0% 0.083%      192MB 15.88%  k8s.io/client-go/rest.TransportFor
    5.36MB  0.44%  0.53%      186MB 15.39%  k8s.io/client-go/transport.(*tlsTransportCache).get
         0     0%  0.53%      186MB 15.39%  k8s.io/client-go/transport.New
   13.50MB  1.12%  1.64%    96.54MB  7.99%  k8s.io/client-go/transport.TLSConfigFor
   69.60MB  5.76%  7.40%    69.60MB  5.76%  k8s.io/client-go/transport.tlsConfigKey
         0     0%  7.40%    54.53MB  4.51%  crypto/tls.X509KeyPair
   34.03MB  2.82% 10.22%    34.03MB  2.82%  encoding/pem.Decode
         0     0% 10.22%    33.01MB  2.73%  crypto/tls.parsePrivateKey
    2.50MB  0.21% 10.42%    33.01MB  2.73%  crypto/x509.ParsePKCS1PrivateKey
         0     0% 10.42%    28.51MB  2.36%  k8s.io/client-go/transport.rootCertPool
   25.01MB  2.07% 12.49%    25.01MB  2.07%  math/big.nat.make (inline)
    2.50MB  0.21% 12.70%    24.01MB  1.99%  crypto/x509.(*CertPool).AppendCertsFromPEM
         0     0% 12.70%       17MB  1.41%  encoding/asn1.Unmarshal (inline)
         0     0% 12.70%       17MB  1.41%  encoding/asn1.UnmarshalWithParams
       4MB  0.33% 13.03%       17MB  1.41%  encoding/asn1.parseBigInt
         0     0% 13.03%       17MB  1.41%  encoding/asn1.parseField
         0     0% 13.03%    14.50MB  1.20%  golang.org/x/net/http2.ConfigureTransports (inline)
      11MB  0.91% 13.94%    14.50MB  1.20%  golang.org/x/net/http2.configureTransports
         0     0% 13.94%    14.50MB  1.20%  k8s.io/apimachinery/pkg/util/net.SetTransportDefaults
         0     0% 13.94%    14.50MB  1.20%  k8s.io/apimachinery/pkg/util/net.configureHTTP2Transport
    1.50MB  0.12% 14.06%    13.50MB  1.12%  crypto/rsa.(*PrivateKey).Precompute
         0     0% 14.06%       13MB  1.08%  math/big.(*Int).SetBytes (inline)
         0     0% 14.06%       13MB  1.08%  math/big.nat.setBytes
         0     0% 14.06%     9.50MB  0.79%  math/big.(*Int).Mod
         0     0% 14.06%     9.50MB  0.79%  math/big.(*Int).QuoRem
         0     0% 14.06%     9.50MB  0.79%  math/big.nat.div
         0     0% 14.06%     9.50MB  0.79%  math/big.nat.divLarge
       9MB  0.74% 14.81%        9MB  0.74%  crypto/x509.(*CertPool).addCertFunc (inline)
       6MB   0.5% 15.31%        6MB   0.5%  k8s.io/client-go/rest.(*Config).TransportConfig
    4.50MB  0.37% 15.68%     4.50MB  0.37%  crypto/x509.NewCertPool (inline)
         0     0% 15.68%     3.50MB  0.29%  golang.org/x/net/http2.registerHTTPSProtocol
    3.50MB  0.29% 15.97%     3.50MB  0.29%  net/http.(*Transport).RegisterProtocol
         0     0% 15.97%     2.50MB  0.21%  math/big.(*Int).ModInverse
         0     0% 15.97%     2.50MB  0.21%  math/big.(*Int).Set (inline)
         0     0% 15.97%     2.50MB  0.21%  math/big.nat.set (inline)
```

This appears to have started with https://github.com/kubernetes/kubernetes/pull/112450/files#diff-4dc8e8f75c3aae18ff37f80d409ed39256f134bdc54f4a42deb81f1694c19fb4R115, since each call of `rest.TransportFor` in `k8s.io/kube-aggregator/pkg/apiserver.(*proxyHandler).updateAPIService` accesses the cache with a unique `DialHolder` pointer.

### What did you expect to happen?

Even with a zero cache hit rate, the cache should not grow indefinitely.

### How can we reproduce it (as minimally and precisely as possible)?

Trigger many APIServiceRegistrationController resyncs by repeatedly patching an APIService (e.g. update an annotation with random bytes). Watch kube-apiserver's heap inuse_space grow.

### Anything else we need to know?

_No response_

### Kubernetes version

<details>

```console
$ kubectl version
Client Version: v1.26.2
Kustomize Version: v4.5.7
Server Version: v1.28.0-alpha.0.29+ed5dd38dcdcdab
```

</details>

### Cloud provider

<details>

</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

### PR DESCRIPTION
**Pull request by enj**
/kind bug
/triage accepted
/assign @liggitt 

cc @benluddy

Fixes #117250

```release-note
Fixed a regression in 1.26 causing a memory leak in the Kubernetes API server that occurs during APIService processing.
```

### CODE REVIEW

File: staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_proxy.go [hunk 1]
```diff
@@ -247,41 +241,32 @@ func (r *proxyHandler) updateAPIService(apiService *apiregistrationv1api.APIServ
...
-		egressDialer, err := r.egressSelector.Lookup(networkContext)
-		if err != nil {
-			klog.Warning(err.Error())
-		} else {
-			newInfo.restConfig.Dial = egressDialer
-		}
-	} else if r.proxyTransport != nil && r.proxyTransport.DialContext != nil {
-		newInfo.restConfig.Dial = r.proxyTransport.DialContext
+	if r.proxyTransportDial != nil {
+		newInfo.transportConfig.DialHolder = r.proxyTransportDial
```
Comment (benluddy) [inline staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_proxy.go:267]: Can this assignment be folded into the `transport.Config` struct literal on L244?

**Review by benluddy — COMMENTED** [review]
/lgtm

Thanks. I ran these changes locally and can confirm I don't see the leak anymore.

File: staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_proxy.go (hunk 1 above)
Comment (enj) [reply staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_proxy.go]: Good catch.  Fixed.

**Comment by aojea** [comment]
I talked with Ben offline, why this didn't leak before?

**Comment by liggitt** [comment]
> I talked with Ben offline, why this didn't leak before?

prior to 1.26 (https://github.com/kubernetes/kubernetes/pull/112450), setting Dial in rest.Config made the resulting client uncacheable

**Comment by aojea** [comment]
> prior to 1.26 (#112450), setting Dial in rest.Config made the resulting client uncacheable

oh, right, so what Ben commented with me is right, if we remove it from the cache we'll have the same situation as it is was uncacheable ... finally I got it

**Pull request merged**